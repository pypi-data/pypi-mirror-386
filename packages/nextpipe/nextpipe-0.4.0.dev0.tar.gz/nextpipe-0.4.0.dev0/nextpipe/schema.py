"""
Schema definitions for Nextpipe.

This module contains schema definitions used for pipeline configurations.

Classes
-------
AppOption
    Option for running an app.
AppRunConfig
    Configuration for running an app.
"""

from dataclasses import dataclass, field
from typing import Any, Optional, Union

from dataclasses_json import dataclass_json

from . import utils


@dataclass_json
@dataclass
class AppOption:
    """
    Option for running an app.

    You can import the `AppOption` class directly from `nextpipe`:

    ```python
    from nextpipe import AppOption
    ```

    This class represents a key-value pair for specifying options when running an app
    in a pipeline.

    Parameters
    ----------
    name : str
        Key for the option.
    value : Any
        Value for the option.

    Examples
    --------
    >>> from nextpipe import AppOption
    >>> option = AppOption(name="threads", value=4)
    """

    name: str
    """Key for the option."""
    value: Any
    """Value for the option."""


@dataclass_json
@dataclass
class AppRunConfig:
    """
    Configuration for running an app.

    You can import the `AppRunConfig` class directly from `nextpipe`:

    ```python
    from nextpipe import AppRunConfig
    ```

    This class represents a configuration object used when running an app
    in a pipeline, containing input data, options, and an optional name.

    Parameters
    ----------
    input : Union[dict[str, Any], str, Any]
        Input data for the app. A JSON app can take a dictionary, multi-file apps can take
        a directory path as a string. Other types will be passed to the underlying Python
        SDK as-is (e.g., nextmv.Input).
    options : Union[list[AppOption], dict[str, Any]], optional
        Options for running the app, by default empty. These can be provided as a list of
        `AppOption` instances, or, simply as a dictionary of key-value pairs.
    name : str, optional
        Name for the run, by default None.

    Examples
    --------
    >>> from nextpipe import AppRunConfig, AppOption
    >>> config = AppRunConfig(
    ...     input={"data": [1, 2, 3]},
    ...     options={"threads": 4},
    ...     name="my-run"
    ... )
    """

    input: Union[dict[str, Any], str, Any]
    """Input for the app. A JSON app can take a dictionary, multi-file apps can take a
    directory path as a string. Other types will be passed to the underlying Python SDK
    as-is (e.g., nextmv.Input)."""
    options: Union[list[AppOption], dict[str, Any]] = field(default_factory=list)
    """Options for running the app."""
    name: Optional[str] = None
    """Name for the run."""

    def get_options(self) -> dict[str, Any]:
        """
        Get options as a dictionary.

        This method converts the `options` attribute to a dictionary if it is provided
        as a list of `AppOption` instances.

        Returns
        -------
        dict[str, Any]
            Dictionary of options.
        """
        options = self.options
        if isinstance(self.options, list):
            options = {option.name: option.value for option in self.options}
        options = utils.convert_to_string_values(options)
        return options
