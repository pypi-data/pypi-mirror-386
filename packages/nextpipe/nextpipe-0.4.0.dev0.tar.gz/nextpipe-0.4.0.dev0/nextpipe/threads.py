"""
Thread management utilities for parallel job execution.

This module provides classes for managing multithreaded job execution.

Classes
-------
Job
    A callable job that can be submitted to a thread pool.
Pool
    A thread pool for executing jobs in parallel.

The module also provides a thread local storage variable for thread-specific data.
"""

import threading
from collections.abc import Callable
from typing import Any, Optional

DEFAULT_MAX_THREADS = 24
"""
Default maximum number of threads for the thread pool. This is set to 24, to allow
parallel execution of sub-runs in the pipeline (as these are typically not CPU-bound). A
lower value more fit for the respective hardware is recommended when running many
CPU-bound steps.
"""

thread_local = threading.local()
"""Thread-local storage.

This variable provides thread-specific storage that can be used to store data
that is specific to a particular thread.
"""


class Job:
    """A callable job that can be submitted to a thread pool.

    This class represents a job that can be executed by a worker thread in a
    thread pool. It wraps a target callable with arguments and callback functions
    that are executed when the job starts and completes.

    Parameters
    ----------
    target : Callable
        The function to be executed by the job.
    start_callback : Callable
        Function to be called when the job starts. Takes the job as a parameter.
    done_callback : Callable
        Function to be called when the job completes. Takes the job as a parameter.
    args : Optional[list], optional
        Arguments to be passed to the target function, by default None.
    name : str, optional
        Name for the job, by default None.
    reference : Any, optional
        Reference to any object associated with this job, by default None.

    Attributes
    ----------
    target : Callable
        The function to be executed by the job.
    start_callback : Callable
        Function to be called when the job starts.
    done_callback : Callable
        Function to be called when the job completes.
    args : Optional[list]
        Arguments to be passed to the target function.
    name : str
        Name for the job.
    reference : Any
        Reference to any object associated with this job.
    done : bool
        Whether the job has completed.
    result : Any
        The result of the job execution.
    error : Exception
        Any exception that occurred during job execution.

    Examples
    --------
    >>> def my_task(x, y):
    ...     return x + y
    >>> def on_start(job):
    ...     print(f"Starting job {job.name}")
    >>> def on_done(job):
    ...     print(f"Job {job.name} completed with result: {job.result}")
    >>> job = Job(
    ...     target=my_task,
    ...     start_callback=on_start,
    ...     done_callback=on_done,
    ...     args=[5, 10],
    ...     name="addition_job"
    ... )
    """

    def __init__(
        self,
        target: Callable,
        start_callback: Callable,
        done_callback: Callable,
        args: Optional[list] = None,
        name: str = None,
        reference: Any = None,
    ):
        self.target = target
        self.start_callback = start_callback
        self.done_callback = done_callback
        self.args = args
        self.name = name
        self.reference = reference
        self.done = False
        self.result = None
        self.error = None

    def run(self):
        """Execute the job's target function.

        Executes the target function with the provided arguments (if any)
        and sets the done flag to True. The result of the execution is
        stored in the result attribute.
        """

        if self.args:
            self.result = self.target(*self.args)
        else:
            self.result = self.target()
        self.done = True


class Pool:
    """A thread pool for executing jobs in parallel.

    This class provides a thread pool for executing jobs in parallel. The number
    of threads is limited by the max_threads parameter. Jobs are queued and
    executed as threads become available.

    Parameters
    ----------
    max_threads : int, optional
        Maximum number of threads to use, by default 0 (uses CPU count).

    Attributes
    ----------
    max_threads : int
        Maximum number of threads in the pool.
    counter : int
        Counter used to assign unique IDs to threads.
    waiting : dict
        Dictionary of jobs waiting to be executed, keyed by thread ID.
    running : dict
        Dictionary of running thread objects, keyed by thread ID.
    error : Exception
        Any exception that occurred during job execution.
    lock : threading.Lock
        Lock for thread synchronization.
    cond : threading.Condition
        Condition variable for thread synchronization.

    Examples
    --------
    >>> def task(x):
    ...     return x * 2
    >>> def on_start(job):
    ...     print(f"Starting job")
    >>> def on_done(job):
    ...     print(f"Result: {job.result}")
    >>> pool = Pool(max_threads=4)
    >>> for i in range(10):
    ...     job = Job(
    ...         target=task,
    ...         start_callback=on_start,
    ...         done_callback=on_done,
    ...         args=[i]
    ...     )
    ...     pool.run(job)
    >>> pool.join()  # Wait for all jobs to finish
    """

    def __init__(self, max_threads: int = 0):
        """Initialize a thread pool.

        Parameters
        ----------
        max_threads : int, optional
            Maximum number of threads to use, by default 0.
            If <= 0, DEFAULT_MAX_THREADS is used.
        """

        if max_threads <= 0:
            max_threads = DEFAULT_MAX_THREADS
        self.max_threads = max_threads
        self.counter = 0  # Used to assign unique IDs to threads
        self.waiting = {}
        self.running = {}
        self.error = None
        self.lock = threading.Lock()
        self.cond = threading.Condition(self.lock)

    def run(self, job: Job) -> None:
        """Submit a job to the thread pool for execution.

        The job is first placed in the waiting queue and then assigned to a worker
        thread when one becomes available. If all threads are busy, the method will
        wait until a thread becomes available.

        Parameters
        ----------
        job : Job
            The job to be executed.

        Notes
        -----
        Jobs are executed in the order they are submitted, but the completion
        order may vary depending on execution time.
        """

        with self.lock:
            self.counter += 1
            thread_id = self.counter
            self.waiting[thread_id] = job

        def worker(job: Job, thread_id: int):
            try:
                if job.start_callback is not None:
                    job.start_callback(job)
                job.run()
            except Exception as e:
                job.error = e
                raise RuntimeError(f"Error in thread {thread_id}: {e}") from e
            finally:
                with self.lock:
                    self.running.pop(thread_id, None)
                    self.cond.notify_all()  # Notify others that a thread is available
                if job.done_callback is not None:
                    job.done_callback(job)

        while True:
            with self.lock:
                if len(self.running) < self.max_threads:
                    # Move job from waiting to running
                    kw_args = {"target": worker, "args": (job, thread_id)}
                    if job.name:
                        kw_args["name"] = job.name
                    thread = threading.Thread(**kw_args)
                    self.running[thread_id] = thread
                    self.waiting.pop(thread_id, None)
                    thread.start()
                    break  # Successfully assigned the job to a thread
                else:
                    self.cond.wait()  # Wait until a thread is available

    def wait_one(self) -> None:
        """
        Wait for one job to finish.

        This method blocks until at least one job completes execution.
        It's useful when you want to process completed jobs as they finish
        without waiting for all jobs to complete.
        """

        with self.lock:
            self.cond.wait()

    def join(self) -> None:
        """
        Wait for all jobs to finish.

        This method blocks until all submitted jobs have completed execution.
        It's useful to ensure all parallelized work is completed before proceeding.

        Examples
        --------
        >>> pool = Pool(max_threads=4)
        >>> # Submit jobs to the pool
        >>> for i in range(10):
        ...     job = Job(...)
        ...     pool.run(job)
        >>> # Wait for all jobs to complete
        >>> pool.join()
        >>> print("All jobs completed")
        """

        with self.cond:
            while self.waiting or self.running:
                self.cond.wait()  # Wait until all jobs are finished
