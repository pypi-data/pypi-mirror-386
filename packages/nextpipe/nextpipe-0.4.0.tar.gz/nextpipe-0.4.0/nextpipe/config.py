"""
Configuration module for the pipeline.

This module provides configuration classes for controlling pipeline behavior.

Classes
-------
Configuration
    Configuration parameters for pipeline execution.
"""

from dataclasses import dataclass

from dataclasses_json import dataclass_json


@dataclass_json
@dataclass
class Configuration:
    """
    Configuration for the pipeline.

    This class encapsulates configuration parameters that control how the
    pipeline executes, including parallelization and input limitations.

    Parameters
    ----------
    thread_count : int, default=0
        Number of threads to use for parallel processing.
        If 0, the number of threads is set to the number of available CPUs.
    max_step_inputs : int, default=50
        Maximum number of inputs to a step.
        This is used to avoid accidental combinatorial explosions due to the Cartesian product
        of inputs used when a step has multiple predecessors which are themselves repeated or
        foreach steps.

    Examples
    --------
    >>> from nextpipe import Configuration
    >>> config = Configuration(thread_count=4, max_step_inputs=100)
    >>> config.thread_count
    4
    """

    thread_count: int = 0
    """
    Number of threads to use for parallel processing.
    If 0, the number of threads is set to the number of available CPUs.
    """
    max_step_inputs: int = 50
    """
    Maximum number of inputs to a step.
    This is used to avoid accidental combinatorial explosions due to the Cartesian product
    of inputs used when a step has multiple predecessors which are themselves repeated or
    foreach steps.
    """
