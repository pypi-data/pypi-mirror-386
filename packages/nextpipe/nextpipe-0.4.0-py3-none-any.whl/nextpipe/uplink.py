"""
Module for communicating with the Nextmv platform.

This module provides functionality to communicate pipeline execution status with the
Nextmv platform. It includes data classes for modeling the pipeline graph state and
a client for updating this state in the platform.

Classes
-------
UplinkConfig
    Configuration for the uplink client.
StepDTO
    Data Transfer Object representing a pipeline step.
NodeDTO
    Data Transfer Object representing a node in the pipeline graph.
FlowDTO
    Data Transfer Object representing a flow graph.
FlowUpdateDTO
    Data Transfer Object for updating a flow in the platform.
UplinkClient
    Client for posting graph and node updates to the platform.

Functions
---------
ExcludeIfNone
    Helper function for dataclasses_json to exclude None fields.
"""

import datetime
import os
import threading
import time
from dataclasses import dataclass, field

from dataclasses_json import config, dataclass_json
from nextmv.cloud import Client

from nextpipe.utils import log_internal

FAILED_UPDATES_THRESHOLD = 10
"""Maximum number of consecutive failed updates before termination."""

UPDATE_INTERVAL = 5
"""Interval in seconds between update attempts."""

MAX_DOCS_LENGTH = 1000
"""Maximum length for documentation strings sent to the platform."""

ENV_APP_ID = "NEXTMV_APP_ID"
"""Environment variable name for the application ID."""

ENV_RUN_ID = "NEXTMV_RUN_ID"
"""Environment variable name for the run ID."""


def ExcludeIfNone(value):
    """
    Determine if a value should be excluded from serialization.

    This function is used as a helper for dataclasses_json to exclude None values
    during serialization.

    Parameters
    ----------
    value : Any
        The value to check.

    Returns
    -------
    bool
        True if the value is None and should be excluded, False otherwise.

    Examples
    --------
    >>> ExcludeIfNone(None)
    True
    >>> ExcludeIfNone("something")
    False
    """

    return value is None


@dataclass
class UplinkConfig:
    """
    Configuration for the uplink client.

    Parameters
    ----------
    application_id : str
        The ID of the application.
    run_id : str
        The ID of the run.
    """

    application_id: str
    """The ID of the application."""
    run_id: str
    """The ID of the run."""


@dataclass_json
@dataclass
class StepDTO:
    """
    Data Transfer Object representing a pipeline step.

    A StepDTO represents a step in a pipeline, with its unique identifier,
    dependencies, documentation, and optional associated application.

    Parameters
    ----------
    id : str
        The ID of the step.
    predecessors : list[str]
        The IDs of the nodes that depend on this node.
    docs : str, optional
        The documentation string of the step.
    app_id : str, optional
        The ID of the app this step represents (if any).
    """

    id: str
    """The ID of the step."""
    predecessors: list[str]
    """The IDs of the nodes that depend on this node."""
    docs: str = field(default=None, metadata=config(exclude=ExcludeIfNone))
    """The doc string of the step."""
    app_id: str = field(default=None, metadata=config(exclude=ExcludeIfNone))
    """The ID of the app this step represents (if any)."""


@dataclass_json
@dataclass
class NodeDTO:
    """
    Data Transfer Object representing a node in the pipeline graph.

    A NodeDTO represents a node in the pipeline execution graph, tracking its
    status, dependencies, and relationships.

    Parameters
    ----------
    id : str
        The ID of the node.
    parent_id : str
        Parent step ID.
    predecessor_ids : list[str]
        Predecessor nodes via their IDs.
    status : str
        Status of the node.
    run_id : str, optional
        ID of the associated run, if any.
    """

    id: str
    """
    The ID of the node.
    """
    parent_id: str
    """
    Parent step.
    """
    predecessor_ids: list[str]
    """
    Predecessor nodes via their IDs.
    """
    status: str
    """
    Status of the node.
    """
    run_id: str = field(default=None, metadata=config(exclude=ExcludeIfNone))
    """
    ID of the associated run, if any.
    """


@dataclass_json
@dataclass
class FlowDTO:
    """
    Data Transfer Object representing a flow graph.

    A FlowDTO represents a flow and more importantly its graph and state,
    including steps and nodes.

    Parameters
    ----------
    steps : list[StepDTO]
        Steps in the flow.
    nodes : list[NodeDTO]
        Nodes and their current state.
    """

    steps: list[StepDTO]
    """
    Steps in the flow.
    """
    nodes: list[NodeDTO]
    """
    Nodes and their current state.
    """


@dataclass_json
@dataclass
class FlowUpdateDTO:
    """
    Data Transfer Object for updating a flow in the platform.

    A FlowUpdateDTO represents a flow in the platform, containing the pipeline graph
    and the timestamp of the update.

    Parameters
    ----------
    pipeline_graph : FlowDTO
        The graph of the pipeline.
    updated_at : str, optional
        Time of the update as an RFC3339 string. Will be set automatically.
    """

    pipeline_graph: FlowDTO
    """
    The graph of the pipeline.
    """
    updated_at: str = None
    """
    Time of the update as an RFC3339 string. Will be set automatically.
    """


class UplinkClient:
    """
    Client for posting graph and node updates to the platform.

    This class provides an interface to communicate with the Nextmv platform,
    posting updates about the pipeline execution status and graph structure.

    Parameters
    ----------
    client : nextmv.cloud.Client
        The Nextmv Cloud client.
    config : UplinkConfig
        The configuration for the uplink client.

    Attributes
    ----------
    config : UplinkConfig
        The configuration for the uplink client.
    inactive : bool
        Whether the client is inactive.
    client : nextmv.cloud.Client
        The Nextmv Cloud client.
    flow : dict or FlowUpdateDTO
        The current flow.
    changed : bool
        Whether the flow has changed and needs to be updated.
    """

    def __init__(self, client: Client, config: UplinkConfig):
        """
        Initialize the UplinkClient.

        Parameters
        ----------
        client : nextmv.cloud.Client
            The Nextmv Cloud client.
        config : UplinkConfig, optional
            The configuration for the uplink client. If None, configuration is
            loaded from environment variables.

        Notes
        -----
        If no application ID or run ID is provided, the client will be marked as
        inactive and will not send any updates to the platform.
        """

        if config is None:
            # Load config from environment
            config = UplinkConfig(
                application_id=os.environ.get(ENV_APP_ID),
                run_id=os.environ.get(ENV_RUN_ID),
            )
        self.config = config
        self.inactive = False
        if not self.config.application_id or not self.config.run_id:
            self.inactive = True
            self.terminated = True
            log_internal("No application ID or run ID found, uplink is inactive.")
        self.client = client
        self._lock = threading.Lock()
        self.flow = {}
        self.changed = False
        self._terminate = False
        self._terminated = False
        self._updates_failed = 0

    def _post_node_update(self):
        """
        Post node updates to the platform.

        This internal method sends the current flow state to the Nextmv platform
        with the current timestamp.

        Raises
        ------
        Exception
            If the request to post the flow update fails.
        """

        # Get RFC3339 timestamp in UTC
        timestamp = datetime.datetime.now(datetime.UTC).isoformat()
        self.flow.updated_at = timestamp
        resp = self.client.request(
            "PUT",
            f"/v1/internal/applications/{self.config.application_id}/runs/{self.config.run_id}/flow",
            payload=self.flow.to_dict(),
        )
        if not resp.ok:
            raise Exception(f"Failed to post flow update: {resp.text}")

    def submit_update(self, flow: FlowUpdateDTO):
        """
        Post the full flow and its state to the platform.

        This method submits the flow state to be updated in the Nextmv platform.
        It truncates documentation strings if they exceed the maximum length.

        Parameters
        ----------
        flow : FlowUpdateDTO
            The flow to update.

        Raises
        ------
        ValueError
            If the flow is not a FlowUpdateDTO instance.
        """

        if self.inactive or self._terminate:
            return
        if not isinstance(flow, FlowUpdateDTO):
            raise ValueError(f"Expected FlowDTO, got {type(flow)}")
        # Truncate docs to a maximum length
        for step in flow.pipeline_graph.steps:
            if step.docs and len(step.docs) > MAX_DOCS_LENGTH:
                step.docs = step.docs[:MAX_DOCS_LENGTH] + "..."
        # Inform the client about the new flow
        with self._lock:
            self.flow = flow
            self.changed = True

    def run_async(self):
        """
        Start the uplink client in a separate thread.

        This method starts the uplink client in a separate thread, which will
        post node updates to the platform until terminated. Updates are sent
        at regular intervals defined by UPDATE_INTERVAL.

        If the client is inactive or already terminated, this method returns
        without starting a new thread.
        """

        if self.inactive or self._terminate:
            return

        def run():
            while not self._terminate:
                # Post update, if any
                if self.changed:
                    with self._lock:
                        try:
                            self._post_node_update()
                            self.changed = False
                        except Exception as e:
                            # Update failed, keep in pending
                            log_internal(f"Failed to post flow update (#{self._updates_failed}): {e}")
                            self._updates_failed += 1
                            if self._updates_failed > FAILED_UPDATES_THRESHOLD:
                                # Too many failed updates, terminate
                                self._terminate = True
                else:
                    self._updates_failed = 0
                # Sleep
                time.sleep(UPDATE_INTERVAL)

            # Signal termination
            self._terminated = True

        threading.Thread(target=run).start()

    def terminate(self):
        """
        Terminate the uplink client gracefully.

        This method stops the uplink client's update thread and sends a final
        update to the platform if there are pending changes. It waits for the
        thread to terminate before returning.

        If the client is inactive, this method returns without taking any action.
        """

        if self.inactive:
            return

        # Terminate the client
        self._terminate = True
        while not self._terminated:
            time.sleep(0.1)

        # Send final update
        if self._updates_failed > 0:
            log_internal(f"Uplink client is terminating (failed updates: {self._updates_failed})")
        if self.changed:
            try:
                self._post_node_update()
            except Exception:
                pass
