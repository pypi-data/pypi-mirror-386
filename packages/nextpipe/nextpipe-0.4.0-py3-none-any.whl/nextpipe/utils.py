"""
Utility functions for nextpipe.

This module provides utility functions used throughout the nextpipe framework
for logging, function wrapping, and AST parsing.

Functions
--------
log
    Log a message with step context.
log_internal
    Log an internal nextpipe message.
wrap_func
    Wrap a function to unpack arguments.
convert_to_string_values
    Convert all dictionary values to strings.
get_ast_root
    Find the AST root of a given object.
"""

import ast
import importlib
import inspect
import sys
import threading
from functools import wraps
from typing import Any

THREAD_NAME_PREFIX = "nextpipe-"
"""str: Prefix for thread names created by nextpipe."""


def __get_step_name() -> str:
    """
    Gets the name of the step currently executing in the calling thread.

    Returns
    -------
    str
        The name of the current step if called from a nextpipe thread,
        or 'main' if called from the main thread.
    """

    if threading.current_thread().name.startswith(THREAD_NAME_PREFIX):
        return threading.current_thread().name[len(THREAD_NAME_PREFIX) :]
    return "main"


def log(message: str) -> None:
    """
    Logs a message using stderr. Furthermore, prepends the name of the calling function if it is a step.

    You can import the `log` function directly from `nextpipe`:

    ```python
    from nextpipe import log
    ```

    This function is useful for debugging and providing execution information during pipeline runs.

    Parameters
    ----------
    message : str
        The message to log.

    Examples
    --------
    >>> from nextpipe import log
    >>> log("Processing data")
    Processing data
    >>> # In a step function, it would show:
    >>> # [step_name] Processing data
    """

    step_name = __get_step_name()
    if step_name:
        print(f"[{step_name}] {message}", file=sys.stderr)
    else:
        print(message, file=sys.stderr)


def log_internal(message: str) -> None:
    """
    Logs a message using stderr.

    This function is used internally by the nextpipe framework to log messages
    with a consistent prefix.

    Parameters
    ----------
    message : str
        The message to log.

    Examples
    --------
    >>> from nextpipe.utils import log_internal
    >>> log_internal("Pipeline initialized")
    [nextpipe] Pipeline initialized
    """

    print(f"[nextpipe] {message}", file=sys.stderr)


def wrap_func(function):
    """
    Wraps the given function in a new function that unpacks the arguments given as a tuple.

    This utility is used internally by nextpipe to pass arguments to functions
    that are executed in threads.

    Parameters
    ----------
    function : callable
        The function to wrap.

    Returns
    -------
    callable
        A new function that unpacks its arguments and calls the original function.

    Examples
    --------
    >>> from nextpipe.utils import wrap_func
    >>> def add(a, b):
    ...     return a + b
    >>> wrapped = wrap_func(add)
    >>> wrapped(([1, 2], {}))
    3
    """

    @wraps(function)
    def func_wrapper(args):
        return function(*args[0], **args[1])

    return func_wrapper


def convert_to_string_values(input_dict: dict[str, Any]) -> dict[str, str]:
    """
    Converts all values of the given dictionary to strings.

    This utility is useful when working with configuration objects where
    string representation of values is required.

    Parameters
    ----------
    input_dict : dict[str, Any]
        The dictionary with values to convert.

    Returns
    -------
    dict[str, str]
        A new dictionary with the same keys but all values converted to strings.

    Examples
    --------
    >>> from nextpipe.utils import convert_to_string_values
    >>> d = {'a': 1, 'b': 2.0, 'c': True}
    >>> convert_to_string_values(d)
    {'a': '1', 'b': '2.0', 'c': 'True'}
    """

    return {key: str(value) for key, value in input_dict.items()}


def __is_running_in_notebook():
    """
    Check if the code is running in a Jupyter notebook.

    This function attempts to detect if the code is running in a Jupyter notebook
    by checking for the presence of IPython with a notebook configuration.

    Returns
    -------
    bool
        True if running in a Jupyter notebook, False otherwise.
    """

    try:
        from IPython import get_ipython

        # Check if the IPython instance is a notebook
        if "IPKernelApp" in get_ipython().config:
            return True
        else:
            return False
    except ImportError:
        return False
    except AttributeError:
        return False


def __get_notebook_ast_root(obj: object) -> ast.ClassDef:
    """
    Find the root AST of the given object in a Jupyter notebook.

    This function searches through the notebook execution history to find
    the cell where the object's class was defined, and returns its AST node.

    Parameters
    ----------
    obj : object
        The object whose class definition AST node to find.

    Returns
    -------
    ast.ClassDef
        The AST node representing the class definition.

    Raises
    ------
    ValueError
        If the class definition cannot be found in the notebook history.
    """

    from IPython import get_ipython

    # Get the current IPython instance
    ipython = get_ipython()

    # Go backwards in the history to find the cell where the object's class was defined.
    for i in range(len(ipython.history_manager.input_hist_parsed) - 1, -1, -1):
        # Parse the code of the cell into an AST.
        tree = ast.parse(ipython.history_manager.input_hist_parsed[i])

        # Find the class definition for the given object.
        for node in ast.walk(tree):
            if isinstance(node, ast.ClassDef) and node.name == obj.__class__.__name__:
                return node

    raise ValueError(f"Could not find AST root for {obj.__class__.__name__} in notebook.")


def __get_normal_ast_root(obj: object) -> ast.ClassDef:
    """
    Find the root AST of the given object.

    This function locates and returns the AST node representing the class definition
    of the provided object by inspecting the source code of its module.

    Parameters
    ----------
    obj : object
        The object whose class definition AST node to find.

    Returns
    -------
    ast.ClassDef
        The AST node representing the class definition.

    Raises
    ------
    IndexError
        If the class definition could not be found in the module.
    """

    module = importlib.import_module(obj.__class__.__module__)
    class_name = obj.__class__.__name__
    tree = ast.parse(inspect.getsource(module)).body
    root = [n for n in tree if isinstance(n, ast.ClassDef) and n.name == class_name][0]
    return root


def get_ast_root(obj: object) -> ast.ClassDef:
    """
    Find the root AST of the given object.

    This function determines whether the code is running in a Jupyter notebook
    or a normal Python module, and calls the appropriate helper function to
    find the AST node representing the class definition of the provided object.

    Parameters
    ----------
    obj : object
        The object whose class definition AST node to find.

    Returns
    -------
    ast.ClassDef
        The AST node representing the class definition.

    Raises
    ------
    ValueError
        If running in a notebook and the class definition cannot be found.
    IndexError
        If not running in a notebook and the class definition cannot be found.

    Examples
    --------
    >>> from nextpipe.utils import get_ast_root
    >>> class Example:
    ...     pass
    >>> obj = Example()
    >>> root = get_ast_root(obj)
    >>> isinstance(root, ast.ClassDef)
    True
    >>> root.name
    'Example'
    """

    if __is_running_in_notebook():
        return __get_notebook_ast_root(obj)
    else:
        return __get_normal_ast_root(obj)
