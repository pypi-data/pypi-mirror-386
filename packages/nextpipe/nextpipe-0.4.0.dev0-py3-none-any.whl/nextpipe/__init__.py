"""Framework for Decision Pipeline modeling and execution."""

from .__about__ import __version__
from .config import Configuration as Configuration
from .decorators import app as app
from .decorators import foreach as foreach
from .decorators import join as join
from .decorators import needs as needs
from .decorators import optional as optional
from .decorators import repeat as repeat
from .decorators import step as step
from .flow import FlowGraph as FlowGraph
from .flow import FlowSpec as FlowSpec
from .schema import AppOption as AppOption
from .schema import AppRunConfig as AppRunConfig
from .utils import log as log

VERSION = __version__
"""The version of the nextpipe package."""
