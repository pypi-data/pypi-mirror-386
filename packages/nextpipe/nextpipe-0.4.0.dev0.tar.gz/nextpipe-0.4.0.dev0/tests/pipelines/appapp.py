import json

import nextmv

from nextpipe import FlowSpec, app, needs, step


# >>> Workflow definition
class Flow(FlowSpec):
    @step
    def prepare(input: dict):
        """Prepares the data."""
        input["prepared"] = True
        return input

    @app(app_id="echo")
    @needs(predecessors=[prepare])
    @step
    def solve1():
        """Runs a model."""
        pass

    @app(app_id="echo")
    @needs(predecessors=[solve1])
    @step
    def solve2():
        """Runs another model."""
        pass

    @needs(predecessors=[solve2])
    @step
    def enhance(result: dict):
        """Enhances the result."""
        output = result["solution"]  # Unwrap the solution
        output["echo"]["data"]["enhanced"] = True
        return output


def main():
    # Load input data
    input = nextmv.load()

    # Run workflow
    flow = Flow("DecisionFlow", input.data)
    flow.run()

    # Write out the result
    print(json.dumps(flow.get_result(flow.enhance)))


if __name__ == "__main__":
    main()
