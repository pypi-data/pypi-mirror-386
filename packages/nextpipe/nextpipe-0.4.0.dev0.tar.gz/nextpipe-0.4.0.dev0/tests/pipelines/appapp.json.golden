{
  "echo": {
    "data": {
      "options": {},
      "solution": {
        "echo": {
          "data": {
            "hello": "world!",
            "prepared": true
          },
          "args": []
        }
      },
      "statistics": {
        "run": {
          "duration": 0.123
        }
      },
      "assets": [],
      "enhanced": true
    },
    "args": []
  }
}
